File: OsmAnd-java/src/main/java/net/osmand/wiki/WikiCoreHelper.java
package net.osmand.wiki;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import net.osmand.PlatformUtil;
import net.osmand.osm.io.NetworkUtils;
import net.osmand.shared.data.KQuadRect;
import net.osmand.shared.wiki.WikiHelper;
import net.osmand.shared.wiki.WikiImage;
import net.osmand.shared.wiki.WikiMetadata;
import net.osmand.util.Algorithms;

import org.apache.commons.logging.Log;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public class WikiCoreHelper {

	public static final boolean USE_OSMAND_WIKI_API = true;
	private static final Log LOG = PlatformUtil.getLog(WikiCoreHelper.class);
	private static final String WIKIMEDIA_API_ENDPOINT = "https://commons.wikimedia.org/w/api.php";
	private static final String WIKIDATA_API_ENDPOINT = "https://www.wikidata.org/w/api.php";
	private static final String WIKIDATA_ACTION = "?action=wbgetclaims&property=P18&entity=";
	private static final String WIKIMEDIA_ACTION = "?action=query&list=categorymembers&cmtitle=";
	private static final String CM_LIMIT = "&cmlimit=100";
	private static final String FORMAT_JSON = "&format=json";
	public static final String WIKIMEDIA_FILE = "File:";
	public static final String WIKIMEDIA_CATEGORY = "Category:";
	private static final int THUMB_SIZE = 480;
	private static final int ICON_SIZE = 64;
	public static final String OSMAND_API_ENDPOINT = "https://osmand.net/api/";
	public static final String OSMAND_SEARCH_ENDPOINT = "https://osmand.net/search/";
	private static final String WIKI_PLACE_ACTION = "wiki_place?";
	private static final String GET_WIKI_DATA_ACTION = "get-wiki-data?";
	private static final int DEPT_CAT_LIMIT = 1;
	private static final List<String> IMAGE_EXTENSIONS = new ArrayList<>(Arrays.asList(".jpeg", ".jpg", ".png", ".gif"));


	public static List<OsmandApiFeatureData> getExploreImageList(KQuadRect mapRect, int zoom, String langs) {
		List<OsmandApiFeatureData> wikiImages = new ArrayList<>();
		StringBuilder url = new StringBuilder();
		String baseApiActionUrl = OSMAND_SEARCH_ENDPOINT + GET_WIKI_DATA_ACTION;
		String northWest = String.format(Locale.US, "%f,%f", mapRect.getTop(), mapRect.getLeft());
		String southEast = String.format(Locale.US, "%f,%f", mapRect.getBottom(), mapRect.getRight());
		url.append(baseApiActionUrl);
		try {
			url.append(String.format(Locale.US, "northWest=%s", URLEncoder.encode(northWest, "UTF-8")));
			url.append("&");
			url.append(String.format(Locale.US, "southEast=%s", URLEncoder.encode(southEast, "UTF-8")));
			url.append("&");
			url.append(String.format(Locale.US, "zoom=%d", zoom));
			url.append("&");
			url.append(String.format(Locale.US, "lang=%s", langs));
			url.append("&");
			url.append(String.format(Locale.US, "filters="));
		} catch (UnsupportedEncodingException e) {
			throw new RuntimeException(e);
		}
		LOG.debug("Download images " + url + " {" + Thread.currentThread().getName() + "}");
		getNearbyImagesOsmAndAPIRequest(url.toString(), wikiImages);
		return wikiImages;
	}

	// wiki_place api call
	public static List<WikiImage> getWikiImageList(Map<String, String> tags, NetworkResponseListener listener) {
		WikiHelper.WikiTagData wikiTagData = WikiHelper.INSTANCE.extractWikiTagData(tags);
		String wikidataId = wikiTagData.getWikidataId();
		String wikiCategory = wikiTagData.getWikiCategory();
		String wikiTitle = wikiTagData.getWikiTitle();
		List<WikiImage> wikiImages = wikiTagData.getWikiImages();
		if (USE_OSMAND_WIKI_API) {
			// article // category
			String url = "";
			String baseApiActionUrl = OSMAND_API_ENDPOINT + WIKI_PLACE_ACTION;
			try {
				if (!Algorithms.isEmpty(wikidataId)) {
					url += baseApiActionUrl + "article=" + URLEncoder.encode(wikidataId, "UTF-8");
				}
				if (!Algorithms.isEmpty(wikiCategory)) {
					url += (url.isEmpty() ? baseApiActionUrl : "&") + "category=" + URLEncoder.encode(wikiCategory, "UTF-8");
				}
				if (!Algorithms.isEmpty(wikiTitle)) {
					url += (url.isEmpty() ? baseApiActionUrl : "&") + "wiki=" + URLEncoder.encode(wikiTitle, "UTF-8");
				}
				if (!url.isEmpty()) {
					url += "&" + "addMetaData=" + URLEncoder.encode("true", "UTF-8");
					getImagesOsmAndAPIRequestV2(url, wikiImages, listener);
				}
			} catch (UnsupportedEncodingException e) {
				LOG.error(e);
			}
		} else {
			if (!Algorithms.isEmpty(wikidataId)) {
				getWikidataImageWikidata(wikidataId, wikiImages);
			}
			if (!Algorithms.isEmpty(wikiCategory)) {
				getWikimediaImageCategory(wikiCategory, wikiImages, 0);
			}
		}

		return wikiImages;
	}

	private static List<WikiImage> getWikimediaImageCategory(String categoryName, List<WikiImage> wikiImages, int depth) {
		String url = WIKIMEDIA_API_ENDPOINT + WIKIMEDIA_ACTION + WIKIMEDIA_CATEGORY + categoryName + CM_LIMIT
				+ FORMAT_JSON;
		WikimediaResponse response = sendWikipediaApiRequest(url, WikimediaResponse.class, false);
		if (response != null) {
			List<String> subCategories = new ArrayList<>();
			for (Categorymember cm : response.query.categorymembers) {
				String memberTitle = cm.title;
				if (memberTitle != null) {
					if (memberTitle.startsWith(WIKIMEDIA_CATEGORY)) {
						subCategories.add(memberTitle);
					} else if (memberTitle.startsWith(WIKIMEDIA_FILE)) {
						WikiHelper.INSTANCE.addFile(wikiImages, memberTitle);
					}
				}
			}
			if (depth < DEPT_CAT_LIMIT) {
				for (String subCategory : subCategories) {
					getWikimediaImageCategory(subCategory, wikiImages, depth + 1);
				}
			}
		}
		return wikiImages;
	}


	protected static List<WikiImage> getWikidataImageWikidata(String wikidataId, List<WikiImage> wikiImages) {
		String url = WIKIDATA_API_ENDPOINT + WIKIDATA_ACTION + wikidataId + FORMAT_JSON;
		WikidataResponse response = sendWikipediaApiRequest(url, WikidataResponse.class, false);
		if (response != null && response.claims != null && response.claims.p18 != null) {
			for (P18 p18 : response.claims.p18) {
				String imageFileName = p18.mainsnak.datavalue.value;
				if (imageFileName != null) {
					WikiImage wikiImage = WikiHelper.INSTANCE.getImageData(imageFileName);
					if (wikiImage != null) {
						wikiImages.add(wikiImage);
					}
				}
			}
		}
		return wikiImages;
	}

	private static OsmandAPIResponseV2 getImagesOsmAndAPIRequestV2(String jsonString) {
		try {
			return new Gson().fromJson(jsonString, OsmandAPIResponseV2.class);
		} catch (JsonSyntaxException e) {
			LOG.error(e.getLocalizedMessage());
		}
		return null;
	}

	private static List<WikiImage> createWikiImages(OsmandAPIResponseV2 response, List<WikiImage> wikiImages) {
		if (response != null && !Algorithms.isEmpty(response.images)) {
			for (Map<String, String> image : response.images) {
				WikiImage wikiImage = parseImageDataWithMetaData(image);
				if (wikiImage != null && isUrlFileImage(wikiImage)) {
					wikiImages.add(wikiImage);
				}
			}
		}
		return wikiImages;
	}

	public static List<WikiImage> getImagesFromJson(String json, List<WikiImage> wikiImages) {
		OsmandAPIResponseV2 response = getImagesOsmAndAPIRequestV2(json);
		return createWikiImages(response, wikiImages);
	}

	private static List<WikiImage> getImagesOsmAndAPIRequestV2(String url, List<WikiImage> wikiImages, NetworkResponseListener listener) {
		OsmandAPIResponseV2 response = sendWikipediaApiRequest(url, OsmandAPIResponseV2.class, listener, false);
		return createWikiImages(response, wikiImages);
	}

	private static boolean isUrlFileImage(WikiImage wikiImage) {
		String path = wikiImage.getImageHiResUrl();
		int lastIndexOfDot = path.lastIndexOf('.');
		if (lastIndexOfDot != -1) {
			return IMAGE_EXTENSIONS.contains(path.substring(lastIndexOfDot).toLowerCase());
		}
		return false;
	}

	private static void getNearbyImagesOsmAndAPIRequest(String url, List<OsmandApiFeatureData> wikiImages) {
		OsmandAPIFeaturesResponse response = sendWikipediaApiRequest(url, OsmandAPIFeaturesResponse.class, true);
		if (response != null && !Algorithms.isEmpty(response.features)) {
			wikiImages.addAll(response.features);
		}
	}

	private static List<WikiImage> getImagesOsmAndAPIRequest(String url, List<WikiImage> wikiImages) {
		OsmandAPIResponse response = sendWikipediaApiRequest(url, OsmandAPIResponse.class, false);
		if (response != null && !Algorithms.isEmpty(response.images)) {
			for (String imageUrl : response.images) {
				if (imageUrl != null) {
					WikiImage wikiImage = parseImageDataFromFile(imageUrl);
					if (wikiImage != null) {
						wikiImages.add(wikiImage);
					}
				}
			}
		}
		return wikiImages;
	}

	private static WikiImage parseImageDataWithMetaData(Map<String, String> image) {
		String imageUrl = image.get("image");
		if (!Algorithms.isEmpty(image)) {
			WikiImage wikiImage = parseImageDataFromFile(imageUrl);
			if (wikiImage != null) {
				WikiMetadata.Metadata metadata = wikiImage.getMetadata();

				String date = image.get("date");
				if (date != null) {
					metadata.setDate(date);
				}
				String author = image.get("author");
				if (date != null) {
					metadata.setAuthor(author);
				}
				String license = image.get("license");
				if (date != null) {
					metadata.setLicense(license);
				}
				long mediaId = Algorithms.parseLongSilently(image.get("mediaId"), -1);
				wikiImage.setMediaId(mediaId);
				return wikiImage;
			}
		}
		return null;
	}

	private static WikiImage parseImageDataFromFile(String imageUrl) {
		String imageHiResUrl = imageUrl.replace("%20", " ").replace(" ", "_");
		try {
			imageUrl = URLDecoder.decode(imageUrl, "UTF-8");
		} catch (IllegalArgumentException | UnsupportedEncodingException e) {
			LOG.error(e.getLocalizedMessage());
		}
		String imageFileName = Algorithms.getFileWithoutDirs(imageUrl);
		String imageName = Algorithms.getFileNameWithoutExtension(imageUrl);
		String imageStubUrl = imageHiResUrl + "?width=" + THUMB_SIZE;
		String imageIconUrl = imageHiResUrl + "?width=" + ICON_SIZE;
		return new WikiImage(imageFileName, imageName, imageStubUrl, imageHiResUrl, imageIconUrl);
	}

	private static <T> T sendWikipediaApiRequest(String url, Class<T> responseClass, boolean useGzip) {
		return sendWikipediaApiRequest(url, responseClass, null, useGzip);
	}

	private static <T> T sendWikipediaApiRequest(String url, Class<T> responseClass, NetworkResponseListener listener, boolean useGzip) {
		StringBuilder rawResponse = new StringBuilder();
		try {
			// Send the GET request with GZIP support
			String errorMessage = NetworkUtils.sendGetRequest(url, null, rawResponse, useGzip);
			if (errorMessage == null) {
				try {
					// Parse the JSON response
					String stringResponse = rawResponse.toString();
					if (listener != null) {
						listener.onGetRawResponse(stringResponse);
					}
					return new Gson().fromJson(stringResponse, responseClass);
				} catch (JsonSyntaxException e) {
					LOG.error(e.getLocalizedMessage());
				}
			} else {
				LOG.error(errorMessage);
			}
		} catch (Exception e) {
			LOG.error(e.getLocalizedMessage());
		}
		return null;
	}

	public static class OsmandAPIFeaturesResponse {
		@SerializedName("features")
		@Expose
		private final List<OsmandApiFeatureData> features = null;
	}

	public static class OsmandAPIResponseV2 {
		@SerializedName("features-v2")
		@Expose
		private final Set<Map<String, String>> images = null;
	}

	public static class OsmandApiFeatureData {
		@Expose
		public WikiDataProperties properties;
		@Expose
		public WikiDataGeometry geometry;
	}

	public static class WikiDataGeometry {
		@Expose
		public double[] coordinates;
	}

	public static class WikiDataProperties {
		public String id;
		public String photoId;
		public String photoTitle;
		public String wikiTitle;
		public String poitype;
		public String poisubtype;
		public String catId;
		public String catTitle;
		public String depId;
		public String depTitle;
		public String wikiLang;
		public String wikiDesc;
		public String wikiLangs;
		public String wikiLangViews;
		public Long osmid;

		public Double elo;
		public int osmtype;
	}

	public static class OsmandAPIResponse {
		@SerializedName("features")
		@Expose
		private final List<String> images = null;
	}

	// Wikidata response classes
	public static class WikidataResponse {
		@SerializedName("claims")
		@Expose
		private Claims claims;
	}

	public static class Claims {
		@SerializedName("P18")
		@Expose
		private final List<P18> p18 = null;
	}

	public static class P18 {
		@SerializedName("mainsnak")
		@Expose
		private Mainsnak mainsnak;
	}

	public static class Mainsnak {
		@SerializedName("datavalue")
		@Expose
		private Datavalue datavalue;
		@SerializedName("datatype")
		@Expose
		private String datatype;
	}

	public static class Datavalue {
		@SerializedName("value")
		@Expose
		private String value;
		@SerializedName("type")
		@Expose
		private String type;
	}

	// Wikimedia response classes
	public static class WikimediaResponse {
		@SerializedName("query")
		@Expose
		private Query query;
	}

	public static class Query {
		@SerializedName("categorymembers")
		@Expose
		private List<Categorymember> categorymembers;
	}

	public static class Categorymember {
		@SerializedName("title")
		@Expose
		private String title;
	}

	public interface NetworkResponseListener {
		void onGetRawResponse(String response);
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/wiki/WikiCoreHelper.kt
package net.osmand.shared.wiki

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import net.osmand.shared.data.KQuadRect
import net.osmand.shared.util.KAlgorithms
import net.osmand.shared.util.LoggerFactory
import net.osmand.shared.util.PlatformUtil
import net.osmand.shared.util.UrlEncoder
import net.osmand.shared.wiki.WikiHelper.addFile

object WikiCoreHelper {

	private val LOG = LoggerFactory.getLogger("WikiCoreHelper")

	const val USE_OSMAND_WIKI_API = true

	private const val WIKIMEDIA_API_ENDPOINT = "https://commons.wikimedia.org/w/api.php"
	private const val WIKIDATA_API_ENDPOINT = "https://www.wikidata.org/w/api.php"
	private const val WIKIDATA_ACTION = "?action=wbgetclaims&property=P18&entity="
	private const val WIKIMEDIA_ACTION = "?action=query&list=categorymembers&cmtitle="
	private const val CM_LIMIT = "&cmlimit=100"
	private const val FORMAT_JSON = "&format=json"

	const val WIKIMEDIA_FILE = "File:"
	const val WIKIMEDIA_CATEGORY = "Category:"
	private const val THUMB_SIZE = 480
	private const val ICON_SIZE = 64

	const val OSMAND_API_ENDPOINT = "https://osmand.net/api/"
	const val OSMAND_SEARCH_ENDPOINT = "https://osmand.net/search/"
	private const val WIKI_PLACE_ACTION = "wiki_place?"
	private const val GET_WIKI_DATA_ACTION = "get-wiki-data?"
	private const val DEPT_CAT_LIMIT = 1

	private val IMAGE_EXTENSIONS = listOf(".jpeg", ".jpg", ".png", ".gif")

	@OptIn(ExperimentalSerializationApi::class)
	private val jsonParser = Json {
		ignoreUnknownKeys = true
		coerceInputValues = true
		explicitNulls = false
		isLenient = true
	}

	fun getExploreImageList(
		mapRect: KQuadRect,
		zoom: Int,
		langs: String
	): List<OsmandApiFeatureData> {
		val wikiImages = mutableListOf<OsmandApiFeatureData>()
		val baseApiActionUrl = "$OSMAND_SEARCH_ENDPOINT$GET_WIKI_DATA_ACTION"

		val nw = "${mapRect.top.toString()},${mapRect.left.toString()}"
		val se = "${mapRect.bottom.toString()},${mapRect.right.toString()}"

		val url = StringBuilder(baseApiActionUrl).apply {
			append("northWest=").append(UrlEncoder.encode(nw))
			append("&southEast=").append(UrlEncoder.encode(se))
			append("&zoom=").append(zoom)
			append("&lang=").append(langs)
			append("&filters=")
		}.toString()

		LOG.debug("Download images $url")
		getNearbyImagesOsmAndAPIRequest(url, wikiImages)
		return wikiImages
	}

	fun getWikiImageList(
		tags: Map<String, String>,
		listener: NetworkResponseListener? = null
	): List<WikiImage> {
		val wikiTagData = WikiHelper.extractWikiTagData(tags)
		val wikidataId = wikiTagData.wikidataId
		val wikiCategory = wikiTagData.wikiCategory
		val wikiTitle = wikiTagData.wikiTitle

		val wikiImages = wikiTagData.wikiImages
		if (USE_OSMAND_WIKI_API) {
			val params = mutableListOf<String>()

			wikidataId?.takeIf { it.isNotEmpty() }?.let {
				params.add("article=${UrlEncoder.encode(it)}")
			}
			wikiCategory?.takeIf { it.isNotEmpty() }?.let {
				params.add("category=${UrlEncoder.encode(it)}")
			}
			wikiTitle?.takeIf { it.isNotEmpty() }?.let {
				params.add("wiki=${UrlEncoder.encode(it)}")
			}
			if (params.isNotEmpty()) {
				val urlParams = params.joinToString("&")
				val finalUrl = "$OSMAND_API_ENDPOINT$WIKI_PLACE_ACTION$urlParams&addMetaData=true"
				getImagesOsmAndAPIRequestV2(finalUrl, wikiImages, listener)
			}
		} else {
			wikidataId?.takeIf { it.isNotEmpty() }?.let { getWikidataImageWikidata(it, wikiImages) }
			wikiCategory?.takeIf { it.isNotEmpty() }
				?.let { getWikimediaImageCategory(it, wikiImages, 0) }
		}
		return wikiImages
	}

	fun getImagesFromJson(json: String, wikiImages: MutableList<WikiImage>): List<WikiImage> {
		return try {
			val response = jsonParser.decodeFromString(OsmandAPIResponseV2.serializer(), json)
			createWikiImages(response, wikiImages)
		} catch (e: Exception) {
			LOG.error("Failed to parse JSON from string", e)
			wikiImages
		}
	}

	private fun getNearbyImagesOsmAndAPIRequest(
		url: String,
		wikiImages: MutableList<OsmandApiFeatureData>
	) {
		val response = sendWikipediaApiRequest<OsmandAPIFeaturesResponse>(url, useGzip = true)
		response?.features?.let { wikiImages.addAll(it) }
	}

	private fun getImagesOsmAndAPIRequestV2(
		url: String,
		wikiImages: MutableList<WikiImage>,
		listener: NetworkResponseListener?
	): List<WikiImage> {
		val response =
			sendWikipediaApiRequest<OsmandAPIResponseV2>(url, useGzip = false, listener = listener)
		return createWikiImages(response, wikiImages)
	}

	private fun createWikiImages(
		response: OsmandAPIResponseV2?,
		wikiImages: MutableList<WikiImage>
	): List<WikiImage> {
		response?.images?.forEach { obj ->
			if (obj is JsonObject) {
				val wikiImage = parseImageDataWithMetaData(obj)
				if (wikiImage != null && isUrlFileImage(wikiImage)) {
					wikiImages.add(wikiImage)
				}
			}
		}
		return wikiImages
	}

	private fun getWikimediaImageCategory(
		categoryName: String,
		wikiImages: MutableList<WikiImage>,
		depth: Int
	): List<WikiImage> {
		val name = if (categoryName.startsWith(WIKIMEDIA_CATEGORY)) categoryName else WIKIMEDIA_CATEGORY + categoryName
		val url = "$WIKIMEDIA_API_ENDPOINT$WIKIMEDIA_ACTION$name$CM_LIMIT$FORMAT_JSON"

		val response = sendWikipediaApiRequest<WikimediaResponse>(url, useGzip = false)
		val members = response?.query?.categorymembers
		if (!members.isNullOrEmpty()) {
			val subCategories = mutableListOf<String>()
			for (cm in members) {
				val memberTitle = cm.title ?: continue
				when {
					memberTitle.startsWith(WIKIMEDIA_CATEGORY) -> subCategories.add(memberTitle)
					memberTitle.startsWith(WIKIMEDIA_FILE) -> addFile(wikiImages, memberTitle)
				}
			}
			if (depth < DEPT_CAT_LIMIT) {
				for (subCategory in subCategories) {
					getWikimediaImageCategory(subCategory, wikiImages, depth + 1)
				}
			}
		}
		return wikiImages
	}

	fun getWikidataImageWikidata(
		wikidataId: String,
		wikiImages: MutableList<WikiImage>
	): List<WikiImage> {
		val url = "$WIKIDATA_API_ENDPOINT$WIKIDATA_ACTION$wikidataId$FORMAT_JSON"
		val response = sendWikipediaApiRequest<WikidataResponse>(url, useGzip = false)

		response?.claims?.p18?.forEach { p18 ->
			p18.mainsnak?.datavalue?.value?.let { imageFileName ->
				val wikiImage = WikiHelper.getImageData(imageFileName)
				wikiImages.add(wikiImage)
			}
		}
		return wikiImages
	}

	private fun parseImageDataWithMetaData(image: JsonObject): WikiImage? {
		val imageUrl = image.getString("image") ?: return null
		val wikiImage = parseImageDataFromFile(imageUrl) ?: return null

		val metadata = wikiImage.metadata
		image.getString("date")?.let { metadata.date = it }
		image.getString("author")?.let { metadata.author = it }
		image.getString("license")?.let { metadata.license = it }

		val mediaIdLong = image.getString("mediaId")?.toLongOrNull() ?: -1L
		wikiImage.setMediaId(mediaIdLong)

		return wikiImage
	}

	private fun parseImageDataFromFile(imageUrl: String): WikiImage? {
		val imageHiResUrl = imageUrl.replace("%20", " ").replace(" ", "_")
		val decodedUrl = UrlEncoder.decode(imageUrl)

		val imageFileName = KAlgorithms.getFileWithoutDirs(decodedUrl)
		val imageName = KAlgorithms.getFileNameWithoutExtension(decodedUrl)

		val imageStubUrl = "$imageHiResUrl?width=$THUMB_SIZE"
		val imageIconUrl = "$imageHiResUrl?width=$ICON_SIZE"

		return WikiImage(imageFileName, imageName, imageStubUrl, imageHiResUrl, imageIconUrl)
	}

	private fun isUrlFileImage(wikiImage: WikiImage): Boolean {
		val path = wikiImage.imageHiResUrl
		val lastDot = path.lastIndexOf('.')
		if (lastDot != -1) {
			val ext = path.substring(lastDot).lowercase()
			return IMAGE_EXTENSIONS.contains(ext)
		}
		return false
	}

	private inline fun <reified T> sendWikipediaApiRequest(
		url: String,
		useGzip: Boolean,
		listener: NetworkResponseListener? = null
	): T? {
		return try {
			val response = PlatformUtil.getNetworkAPI().sendGetRequest(url, null, useGzip)
			if (response.error == null && response.response != null) {
				listener?.onGetRawResponse(response.response)
				jsonParser.decodeFromString<T>(response.response)
			} else {
				LOG.error("Request failed: ${response.error}")
				null
			}
		} catch (e: Exception) {
			LOG.error(e.message ?: "Request/parse failed", e)
			null
		}
	}

	private fun JsonObject.getString(key: String): String? {
		val value = this[key] ?: return null
		return when (value) {
			is JsonNull -> null
			is JsonPrimitive -> value.contentOrNull
			else -> value.toString()
		}
	}

	@Serializable
	data class OsmandAPIFeaturesResponse(val features: List<OsmandApiFeatureData>? = null)

	@Serializable
	data class OsmandAPIResponseV2(@SerialName("features-v2") val images: JsonArray? = null)

	@Serializable
	data class OsmandApiFeatureData(
		val properties: WikiDataProperties? = null,
		val geometry: WikiDataGeometry? = null
	)

	@Serializable
	data class WikiDataGeometry(val coordinates: DoubleArray? = null)

	@Serializable
	data class WikiDataProperties(
		val id: String? = null,
		val photoId: String? = null,
		val photoTitle: String? = null,
		val wikiTitle: String? = null,
		val poitype: String? = null,
		val poisubtype: String? = null,
		val catId: String? = null,
		val catTitle: String? = null,
		val depId: String? = null,
		val depTitle: String? = null,
		val wikiLang: String? = null,
		val wikiDesc: String? = null,
		val wikiLangs: String? = null,
		val wikiLangViews: String? = null,
		val osmid: Long? = null,
		val elo: Double? = null,
		val osmtype: Int = 0
	)

	@Serializable
	data class WikidataResponse(val claims: Claims? = null)

	@Serializable
	data class Claims(@SerialName("P18") val p18: List<P18>? = null)

	@Serializable
	data class P18(val mainsnak: Mainsnak? = null)

	@Serializable
	data class Mainsnak(val datavalue: Datavalue? = null, val datatype: String? = null)

	@Serializable
	data class Datavalue(val value: String? = null, val type: String? = null)

	@Serializable
	data class WikimediaResponse(val query: Query? = null)

	@Serializable
	data class Query(val categorymembers: List<Categorymember>? = null)

	@Serializable
	data class Categorymember(val title: String? = null)

	interface NetworkResponseListener {
		fun onGetRawResponse(response: String)
	}
}
